### ISSUE TITLE
[BUG] Zome calls routed to wrong hApp

### ISSUE THREAD

**Issue by mattyg** [holochain/holochain#2145]
**Describe the bug**
There seem to be a few cases where zome calls are routed to cells in the wrong hApp, if the hApps are distinguished only by app id and coordinator zome changes.

**Expected behavior**
Zome calls via an AppAgentWebsocket should always be routed to cells in that app.

**System information:**
 - OS: Ubuntu 22.04
 - Browser: n/a
 - Holochain and HDK Version: 0.1.3

**Steps to reproduce**
Minimum reproduction: https://github.com/mattyg/bug-reproduction-cell-routing
See README for steps to run

**Screenshots**
n/a

**Additional context**
I wasn't sure if this was be a launcher or holochain core issue, but Matthias looked at it and said it looks like a core bug: https://github.com/holochain/launcher-tauri/issues/150

**Comment by matthme** [comment]
Yes, I confirm that from what I can see it looks like a bug in holochain. Maybe when making zome calls, which coordinator zomes are to be called is only derived from the dna hash but not the complete cell id, i.e. agent public key. So if there are two cells with the same dna hash, it seems to occur that zome calls are being routed to a cell with the same dna hash but belonging to another public key.

**Comment by ThetaSinner** [comment]
Thank you for the report and the reproduction @mattyg. I'm digging into what's going on here and will get back to you when I have something

**Comment by ThetaSinner** [comment]
I can confirm this is a problem with `holochain`. The problem arises because the two apps `example1` and `example2` end up with the same DNA hash. In itself, that isn't a problem, the two `dna.yaml` files are identical so that is expected.

The two applications are recognised as separate everywhere that I've looked so far because the agent id makes it clear that their two cells are distinct.

The problem is that the [zome store](https://github.com/holochain/holochain/blob/develop/crates/holochain/src/conductor/ribosome_store.rs#L10) assumes that the DNA hash is unique. Which in thise case of course, it is not. 

I think this is just a case of a problem assumption being made in this data store.

We will figure out a fix for this and get back to you. As a workaround for now, you can ensure that your DNA files are different to anything else that is running on your `holochain`/conductor instance which will avoid the problem. For example you could set a network seed per app in the `dna.yaml` or when you install them.

**Comment by ThetaSinner** [comment]
@mattyg I have a better answer for you now. This is still about a repeated hash value and I still think that holochain should try and detect this to prevent accidents.
However, there is a component to the hash that I'd missed. We include the hash of the integrity zome wasm but NOT the coordinator zome wasm as part of the DNA hash.

This would also explain the original issue raised against the launcher. If you're trying to install a newer/older version of an app where only the coordinator zome has changed then you're going to have this same problem with holochain not understanding which version is which.

**Comment by mattyg** [comment]
> @mattyg I hdave a better answer for you now. This is still about a repeated hash value and I still think that holochain should try and detect this to prevent accidents. However, there is a component to the hash that I'd missed. We include the hash of the integrity zome wasm but NOT the coordinator zome wasm as part of the DNA hash.
> 
> This would also explain the original issue raised against the launcher. If you're trying to install a newer/older version of an app where only the coordinator zome has changed then you're going to have this same problem with holochain not understanding which version is which.

Yep that's the idea! I'm guessing the necessary change to holochain core might be for the ribosome store to use the Cell Id as it's index key (where the cell id is a tuple of [DnaHash, AgentPubKey]) because the AgentPubKey will be different for every hApp. But I'm not really familiar with the holochain core codebase so I"m not sure if the distinction belongs in the ribosome store or at a different level. 

The backstory for this issue is that I want to publish an update to a hApp that specifically does not change the integrity zomes, and thus doesn't change the DNA hash so I can avoid forking the network.

**Comment by maackle** [comment]
> the AgentPubKey will be different for every hApp

With default settings yes, but an agent key can be specified when installing, meaning you can have two apps with the same agent key. (not that this is a good idea).

But yes, this definitely feels like confusion leftover from when zomes got split into integrity and coordination flavors. The fact that we include coordination zomes in a DNA is confusing, because any set of coordination zomes would lead to the same DNA hash. The fact that we include coordination zomes in DNAs leads us to believe that coordination zomes are in some way tied up with the identity of the DNA, which they're not at all.

So perhaps we have to further this distinction and say that integrity zomes are bound to a **DNA**, and coordinator zomes are bound to a **(DNA + App)**.

> I'm guessing the necessary change to holochain core might be for the ribosome store to use the Cell Id as it's index key

This feels on the right track. This would work if indeed every app had a unique agent, since then the agent key would be a sort of proxy for the app. Since that's not strictly the case though I think we may instead have to use `(DnaHash, InstalledAppId)`. I haven't looked at the code yet, I'm just thinking somewhat quickly and roughly in a spare moment. I'm also not suggesting a particular solution but this seems like the right direction. 

It might have some unfortunate consequences that we'll then have to deal with though. If nothing else, we'll have to communicate and think about coordination zomes a bit more clearly. For instance, the coordinators included in a DNA are not *part of* the DNA, but rather "the initial set of coordinators that will be associated with the Cell that gets created when installing an app using this DNA file".

**Comment by mattyg** [comment]
Yeah the whole coordinator <-> DNA relationship does feel pretty odd and confusing right now. I guess this is partly due to using the biology analogies for naming, and now after refactoring, the analogies have become a bit mismatched.

To me it feels like this is the "accurate" architecture description:

A hApp is:
`[ [ Cell ] <- GUI`

A Cell is:
`[ Dna ] <- Coordinator`

A DNA is:
`[ Integrity Zome ], properties, network seed`

So all together it looks like this:
`[ [ Dna( [ Integrity Zome ], properties, network seed ) ] <-Coordinator ] <- GUI`

In which case it feels kindof weird to even call the coordinators "zomes", since they aren't contained within the DNA. In the biology metaphors they're more like "receptors" + "cellular fluid" (i.e. this diagram stolen from thomascal on discord: https://cdn.discordapp.com/attachments/783168949483077712/1062772980079411211/aW5nLmdpZg.png). 

But honestly I'm afraid this is opening up a can of worms that's best left closed at this stage. I don't know the lightest-touch way to clarify the underlying architecture, which is really just important for newly incoming developers when starting to learn the framework. Maybe you're right that clarifying that a DNA's coordinator zomes as simply the "default coordinator zomes" is a good light-touch approach.

**Comment by maackle** [comment]
> I guess this is partly due to using the biology analogies for naming, and now after refactoring, the analogies have become a bit mismatched.

I agree completely. Once we split zomes into the two types, we wound up with mixed metaphors.

I also agree with your description of the architecture, it looks accurate. In particular, coordinators are not really part of the DNA. FWIW, I spent some time last weekend trying to draw what Holochain looks like to me, just as a personal project, and my depiction of Cells is like how you described: 

![image](https://user-images.githubusercontent.com/896230/229325648-3cf4bd53-8a43-4dbb-adec-186475ff8706.png)

> I don't know the lightest-touch way to clarify the underlying architecture -- though I do feel like getting this right is really important for bringing in new developers to quickly learn the framework.

Yes to this too. We need to change something, but I'm not sure what. Coordinators are "zomes" in the sense of being composable wasm modules, but they've lost the bio meaning of "chromosome". Maybe we need to call them something other than "zomes".

> I'm afraid this is opening up a can of worms that's best left closed at this stage

Yes to this too :upside_down_face: but it is significant that this bug was likely caused by confusing naming, which led to a confused mental model (still stuck in the past where zomes are all part of DNA), which led to a confused implementation.

**Comment by maackle** [comment]
We could start a [Discussion](https://github.com/holochain/holochain/discussions/new/choose) around this, I don't have time to type it up right now but might next week, if you don't before I do.

**Comment by mattyg** [comment]
> We could start a [Discussion](https://github.com/holochain/holochain/discussions/new/choose) around this, I don't have time to type it up right now but might next week, if you don't before I do.

started here: https://github.com/holochain/holochain/discussions/2167

**Comment by abe-njama** [comment]
This issue should be undertaken alongside [updating coordinators](https://hackmd.io/Fh2zdMNSQ6KbAB-dtaiZkA?view)

**Comment by mattyg** [comment]
Bump that this is still a critical issue (for launcher context especially). I think it can be addressed without bundling in an larger "updating coordinators" epic. I originally saw it in launcher when running two instances of the same app where one had updated coordinators.

**Comment by matthme** [comment]
I just gave it a go at trying to reproduce this with holochain 0.4 and I was unable to do so.
I cloned the [reproduction repo](https://github.com/mattyg/bug-reproduction-cell-routing/tree/main), bumped it to hdk 0.4.0-rc.0 / hdi 0.5.0-rc.0 and followed the readme using launcher [0.400.0-rc.2](https://github.com/holochain/launcher/releases/tag/v0.400.0-rc.2). Zome calls seem to be routed fine.
I did see that some of the code around the zome call logic in the conductor has been modified in the meantime so I'm wondering whether it got resolved without noticing.

**Comment by matthme** [comment]
Here is my holochain-0.4 branch of the reproduction repo: https://github.com/matthme/bug-reproduction-cell-routing/tree/holochain-0.4

**Comment by ThetaSinner** [comment]
Should figure out how to solve this as part of #3523

**Comment by matthme** [comment]
@mjbrisebois encountered a possibly related issue recently on holochain 0.4:

>  I'm installing a new happ that has a new coordinator zome, but the WASM hash is the same as the older installation (both are running at the same time)

> As long as the DNA hash remains the same, the DNA definition will always use the first coordinators that were uploaded.

> My guess is that the install process uses the coordinators from the DNA definition rather than the coordinators uploaded in the bundle.

**Comment by ThetaSinner** [comment]
Yep, that sounds related. A side-by-side install should replicate this problem.

**Comment by ThetaSinner** [comment]
> encountered a possibly related issue recently on holochain 0.4:

Split this into another issue, @ThetaSinner

**Comment by jost-s** [comment]
Propopsed solution:

Looking up a zome function to call is performed by DNA hash. When two agents install the same DNA in the conductor, looking up the correct zome function is no longer guaranteed. To fix this, the zome map needs to be indexed by cell ID instead of DNA hash. Zome calls specify the cell ID to identify which zome should be addressed for the call, so using this as the index of the map is logical.

**Comment by matthme** [comment]
I looked quite thoroughly into this now. The underlying problem is that both the ribosome store (as @mattyg already mentioned) as well as the `DnaDef` sqlite table are indexed by dna hash. However, a `DnaDef` contains both integrity zomes *and* coordinator zomes (and should probably not be called `DnaDef` anymore if we understand the "dna" as being the thing that determines the dna hash, i.e. integrity zomes + modifiers).
So we end up only ever being able to store a single set of coordinators for a given dna hash (a.k.a. a given set of integrity zomes + modifiers).
This leads to the following problems:
- if I update coordinators for an existing app, they get updated in-memory in the RibosomeStore but [not persisted in the sqlite database](https://github.com/holochain/holochain/blob/b5fef9e371456f8a2181a023dd144712c9e0bf07/crates/holochain/src/conductor/conductor.rs#L723) since a `DnaFile` already exists for the same dna hash. This can be reproduced by adding
```
    conductor.shutdown().await;
    conductor.startup().await;
```
to the update coordinators test [here](https://github.com/holochain/holochain/blob/772463691eb62d8a39e05bc3dd07cf98800cbe7c/crates/holochain/tests/tests/integrity_zome/mod.rs#L66).
- if I install an app with the same set of [integrity zomes + modifiers] as an already installed app but with different coordinator zomes (and a different agent key), then this app will end up using the same set of coordinator zomes as the app that's already installed, again because a `DnaDef` already exists for that dna hash in the `DnaDef` sqlite table

To solve this, we will want to index both the ribosome store as well as the `DnaDef` table with an index that is
- constant over the lifetime of an app (i.e. it doesn't change when I update coordinator zomes for an app)
- does not lead to collisions for cells with the same set of [integrity zomes + modifiers]

In the long term we will probably want to use tuple `(InstalledAppId, RoleName)` as this identifier which would require a more significant refactor. However, as a quicker fix for the short term, we propose to use `CellId` as this identifier (as @mattyg already proposed). Since we cannot install the same set of [[integrity zomes + modifiers] + coordinator  zomes] twice for the same agent (or else we get a `CellAlreadyExists` error) this is a workable identifier for now that satisfies above two constraints.

This does imply that, if we were to intend to keep the `UseExisting` provisioning strategy, we would be constrained to only ever using one set of coordinator zomes for a given cell id (a.k.a [DnaHash + AgentPubKey]). But this seems acceptable for the time being.

All in all, the proposed changes are:
- change the RibosomeStore to be indexed by CellId instead of DnaHash
- change the DnaDef sqlite table to be indexed by CellId (as a blob) instead of by DnaHash
- some of the methods/functions interacting with the RibosomeStore will need an additional argument cell_id since the identifier cannot anymore be derived from the `DnaDef` alone (the latter is derived from the dna manifest and has no awareness about the agent key with which a cell is going to be installed)
- remove the ability to install a dna (as part of an app) by virtue of specifying an `installed_hash` in the app manifest since we cannot unambiguously infer (and never could since the integrity/coordinator split) from a dna hash which set of [[integrity zomes + modifiers] + coordinator  zomes] is associated to it

**Comment by matthme** [comment]
Fixed with https://github.com/holochain/holochain/pull/5189
